dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -0,0 +1,586 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Common/Arena.h>
+#include <DataStreams/WindowBlockInputStream.h>
+#include <Interpreters/ExpressionActions.h>
+#include <Interpreters/convertFieldToType.h>
+
+namespace DB
+{
+namespace ErrorCodes
+{
+extern const int BAD_ARGUMENTS;
+extern const int NOT_IMPLEMENTED;
+} // namespace ErrorCodes
+
+WindowBlockInputStream::WindowBlockInputStream(const BlockInputStreamPtr & input, const WindowDescription & window_description_, const String & req_id)
+    : log(Logger::get(NAME, req_id))
+    , window_description(window_description_)
+{
+    children.push_back(input);
+    output_header = input->getHeader();
+    for (const auto & add_column : window_description_.add_columns)
+    {
+        output_header.insert({add_column.type, add_column.name});
+    }
+
+    initialWorkspaces();
+
+    initialPartitionAndOrderColumnIndices();
+}
+
+
+void WindowBlockInputStream::initialPartitionAndOrderColumnIndices()
+{
+    partition_column_indices.reserve(window_description.partition_by.size());
+    for (const auto & column : window_description.partition_by)
+    {
+        partition_column_indices.push_back(
+            output_header.getPositionByName(column.column_name));
+    }
+
+    order_column_indices.reserve(window_description.order_by.size());
+    for (const auto & column : window_description.order_by)
+    {
+        order_column_indices.push_back(
+            output_header.getPositionByName(column.column_name));
+    }
+}
+
+void WindowBlockInputStream::initialWorkspaces()
+{
+    // Initialize window function workspaces.
+    workspaces.reserve(window_description.window_functions_descriptions.size());
+
+    for (const auto & window_function_description : window_description.window_functions_descriptions)
+    {
+        WindowFunctionWorkspace workspace;
+        workspace.window_function = window_function_description.window_function;
+        workspace.column_name = window_function_description.column_name;
+        workspace.argument_columns.assign(window_function_description.argument_names.size(), nullptr);
+        workspaces.push_back(std::move(workspace));
+    }
+    only_have_row_number = onlyHaveRowNumber();
+    only_have_pure_window = onlyHaveRowNumberAndRank();
+}
+
+Block WindowBlockInputStream::readImpl()
+{
+    const auto & stream = children.back();
+    while (!input_is_finished)
+    {
+        if (Block output_block = tryGetOutputBlock())
+        {
+            return output_block;
+        }
+
+        Block block = stream->read();
+        if (!block)
+            input_is_finished = true;
+        else
+            appendBlock(block);
+        tryCalculate();
+    }
+
+    if (isCancelledOrThrowIfKilled())
```

Comment:
Why not move this check inside `tryGetOutputBlock`?

Reply:
> Why not move this check inside `tryGetOutputBlock`?

done